lib/srv/db/sqlserver/kinit/kinit.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2022 Gravitational, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package kinit provides utilities for interacting with a KDC (Key Distribution Center) for Kerberos5, or krb5, to allow
+// teleport to connect to sqlserver using x509 certificates.
+package kinit
+
+import (
+	"bytes"
+	"context"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/auth/windows"
+)
+
+const (
+	// krb5ConfigEnv sets the location from which kinit will attempt to read a configuration value
+	krb5ConfigEnv = "KRB5_CONFIG"
+)
+
+// ProviderI is a provider interface for a kinit context
+type ProviderI interface {
```

Comment:
I generally find in Go that it is an anti-pattern to define the interface where the implementation is. (and the appending of `I` makes me think of Java - I've not seen this much in Go)

[To quote the great Dave Cheney:](https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_let_callers_define_the_interface_they_require)

> Interfaces declare the behaviour the caller requires not the behaviour the type will provide. Let callers define an interface that describes the behaviour they expect. The interface belongs to them, the consumer, not you.

Where possible, we should define interfaces at the consumer, and specify exactly what the consumer requires and nothing more. This provides a little more flexibility, and, has the additional benefit of making it clearer what aspects of something a consumer requires. It also prevents the never-ending inflation of an interface.